### PR TITLE
Add support for Sysprep in the API

### PR DESCRIPTION
**Pull request by borod108**
Add support for Sysprep in the API
    
Currently using sysprep is already possible by mounting a configMap
as a cdrom.
This PR allows adding a "sysprep" type volume using either configMap
of Secret so that it is clearly supported in the API.

So now we can use:
```
...
disks:
  - name: sysprep
        cdrom:
        bus: sata
...
volumes:
  - name: sysprep
      sysprep:
         configMap:
           name: sysprep-config
```
(Same with Secret)

Where the configMap is of the form:
```
apiVersion: v1                                                                                                                                                                                              
kind: ConfigMap                                                                                                                                                                                                     
metadata:      
  name: sysprep-config
data:                 
  Unattend.xml: |
     ........
  Autounattend.xml: |
     .......
```

Or secret:
```
apiVersion: v1
kind: Secret
metadata:
  name: sysprep-config
stringData: 
data:                 
  Unattend.xml: |
     ........
  Autounattend.xml: |
     .......
```

```release-note
Added support for Sysprep in the API. A user can now add a answer file through a ConfigMap or a Secret. The User Guide is updated accordingly. /kind feature
```

P.S. Thank you @cben for all the pairing :))

### CODE REVIEW

File: api/api-rule-violations-known.list [hunk 1]
```diff
@@ -245,6 +245,8 @@ API rule violation: names_match,kubevirt.io/client-go/api/v1,LunTarget,ReadOnly
 API rule violation: names_match,kubevirt.io/client-go/api/v1,NetworkConfiguration,NetworkInterface
 API rule violation: names_match,kubevirt.io/client-go/api/v1,PITTimer,Enabled
 API rule violation: names_match,kubevirt.io/client-go/api/v1,RTCTimer,Enabled
+API rule violation: names_match,kubevirt.io/client-go/api/v1,SysprepSource,UserDataConfigMap
+API rule violation: names_match,kubevirt.io/client-go/api/v1,SysprepSource,UserDataSecretRef
```
Comment (cben) [inline api/api-rule-violations-known.list:249]: I think these will go away if you drop "UserData" prefix from Go struct field names.

We copied this prefix from cloud-init code (which also has `NetworkDataSecretRef` etc.) but it doesn't really make sense for Sysprep.
(We = Boris and I pair-programmed on some of this PR)

File: pkg/config/config.go [hunk 2]
```diff
@@ -52,6 +52,8 @@ const (
 var (
 	// ConfigMapSourceDir represents a location where ConfigMap is attached to the pod
 	ConfigMapSourceDir = mountBaseDir + "/config-map"
+	// SysprepSourceDir represents a location where a Sysprep ConfigMap is attached to the pod
```
Comment (cben) [inline pkg/config/config.go:55]: Drop "ConfigMap" from this comment, it applies to both ConfigMap & Secret cases.
Same below in `SysprepDisksDir` comment.

File: staging/src/kubevirt.io/client-go/api/v1/schema.go [hunk 3]
```diff
@@ -599,6 +611,9 @@ type VolumeSource struct {
 	// More info: https://cloudinit.readthedocs.io/en/latest/topics/datasources/configdrive.html
 	// +optional
 	CloudInitConfigDrive *CloudInitConfigDriveSource `json:"cloudInitConfigDrive,omitempty"`
+	// Represents a Sysprep user data source.
```
Comment (cben) [inline staging/src/kubevirt.io/client-go/api/v1/schema.go:614]:
```suggestion
 	// Represents a Sysprep volume source.
```

File: api/api-rule-violations-known.list (hunk 1 above)
Comment (borod108) [reply api/api-rule-violations-known.list]: Right :)

**Comment by borod108** [comment]
@omeryahud  would you care to take a look?

**Comment by borod108** [comment]
@omeryahud @ksimon1 @vatsalparekh care to review?

File: pkg/config/sysprep.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+func GetSysprepSourcePath(volumeName string) string {
+	return filepath.Join(SysprepSourceDir, volumeName)
+}
+
+// GetSysprepDiskPath returns a path to a ConfigMap iso image created based on a volume name
+func GetSysprepDiskPath(volumeName string) string {
+	return filepath.Join(SysprepDisksDir, volumeName+".iso")
+}
+
+func sysprepVolumeHasContenets(sysprepVolume v1.SysprepSource) bool {
```
Comment (dhiller) [inline pkg/config/sysprep.go:45]: nit: spelling
```suggestion
func sysprepVolumeHasContents(sysprepVolume v1.SysprepSource) bool {
```

File: pkg/config/sysprep.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+	}
+
+	return fmt.Errorf("Sysprep drive should contain %s, but it was not found", sysprepFileName)
+}
+
+// CreateSysprepDisks creates Sysprep iso disks which are attached to vmis from either ConfigMap or Secret as a source
+func CreateSysprepDisks(vmi *v1.VirtualMachineInstance) error {
+	for _, volume := range vmi.Spec.Volumes {
+		if volume.Sysprep != nil {
+			if sysprepVolumeHasContenets(*volume.Sysprep) {
```
Comment (dhiller) [inline pkg/config/sysprep.go:70]: nit: spelling
```suggestion
			if sysprepVolumeHasContents(*volume.Sysprep) {
```

**Review by dhiller — COMMENTED** [review]
Hey @borod108 , thanks for your work! I've only found one nit regarding spelling, otherwise this looks fine!

File: pkg/config/sysprep.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (borod108) [reply pkg/config/sysprep.go:45]: @dhiller thank you for catching this :)

**Comment by dhiller** [comment]
@borod108 thanks for fixing the issue!

Remaining todos are
* [fixing the unit tests](https://prow.apps.ovirt.org/view/gcs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/4921/pull-kubevirt-unit-test/1359245487475003392) and 
* [adding a release note](https://github.com/kubevirt/kubevirt/pull/4921#issuecomment-769122460) (I figure it's worth it for your change :) )

**Review by dhiller — APPROVED** [review]
/lgtm
/approve

**Comment by rmohr** [comment]
I may be mistaken, but I don't see e2e tests for this.

**Comment by borod108** [comment]
@rmohr I will look into it. And wanted to say thank you for all the help you have provided with understanding what is to be done this PR!

**Pull request merged**